### ISSUE TITLE
Question regarding implementation of the variable out_of_bound

### ISSUE THREAD

**Issue by asprasan** [ClementPinard/SfmLearner-Pytorch#38]
Consider the following line:
`            out_of_bound = 1 - (ref_img_warped == 0).prod(1, keepdim=True).type_as(ref_img_warped)`
in the function `loss_functions.py`. If I am not wrong this is computed to not take loss at those pixels which are out of bound or possibly, disoccluded as well due to warping. My question here is, wouldn't pytorch calculate the gradient with respect to this variable `out_of_bound` as well and update it, although it shouldn't be updated?

This may not change the overall code and performance, but still a correct way to do it would be to specify that the variable `out _of_bound` should not get any gradients and not be updated as well.

**Comment by ClementPinard** [comment]
the `ref_img_warped == 0` variable is not differentiable since it's binary operation (it's either 0 or 1, so it's either constant or not even continuous).
So the gradients are not propagated through this variable. However, you're right that the gradient can propagate at least until this binary tensor, while we already know it's a dead en from `out_of_bound`, so a few computation is wasted, but I'd say that the overhead is very low.

To make sure you don't compute any gradient wrt to `out_of_bound`, you can use the context manager `with torch.no_grad():` before setting this variable, but I doubt that it will be very useful.

**Comment by asprasan** [comment]
Yes. That makes sense. It doesn't put too much overhead. 

Thank you very much.
